<a href="https://colab.research.google.com/github/JaswanthJavangula/ML_Practise/blob/main/1ML%20JAshu/NAIVE%20bayes/prac_GNB_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import load_wine
import pandas as pd

df= pd.DataFrame(load_wine().data,columns=load_wine().feature_names)
df['target']=load_wine().target
print(df.isnull().sum())
df.info()


alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
target                          0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    f

In [ ]:
df.head(3)

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0,0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0,0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0,0


In [ ]:
X = df.drop("target", axis= 1)
y = df.target

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score,confusion_matrix
#TTS
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35, random_state=42,stratify =y)
#Intiallizing GNB
gnb = GaussianNB()
gnb.fit(X_train,y_train)
y_pred = gnb.predict(X_test)
#SCORES
print(accuracy_score(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))


0.9523809523809523
[[19  2  0]
 [ 1 24  0]
 [ 0  0 17]]


removing couple features


In [ ]:
df_copy = df.drop(["ash",	"alcalinity_of_ash",	"magnesium",	"hue","od280/od315_of_diluted_wines","proline"],axis=1)

In [ ]:
X1 = df_copy.drop("target", axis= 1)
y1 = df_copy.target
X1.shape, y1.shape

((178, 7), (178,))

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score,confusion_matrix
#TTS
X_train1, X_test1, y_train1, y_test1 = train_test_split(X1, y1, test_size=0.35, random_state=42,stratify =y)
#Intiallizing GNB
gnb1 = GaussianNB()
gnb1.fit(X_train1,y_train1)
y_pred1 = gnb1.predict(X_test1)
#SCORES
print(accuracy_score(y_test1,y_pred1))
print(confusion_matrix(y_test1,y_pred1))


0.873015873015873
[[18  3  0]
 [ 3 21  1]
 [ 0  1 16]]


STANDARD SCALER

In [ ]:
df_copy1 = df_copy.drop(["total_phenols"],axis=1)
X2 = df_copy1.drop("target", axis= 1)
y2 = df_copy1.target
X2.shape, y2.shape

((178, 6), (178,))

In [ ]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size=0.35, random_state=42,stratify =y)

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train2)
X_test_scaled = scaler.transform(X_test2)

gnb_scaled = GaussianNB()
gnb_scaled.fit(X_train_scaled, y_train2)
y_pred_scaled = gnb_scaled.predict(X_test_scaled)
#SCORES
print(accuracy_score(y_test2,y_pred_scaled))
print(confusion_matrix(y_test2,y_pred_scaled))

0.8888888888888888
[[18  3  0]
 [ 2 22  1]
 [ 0  1 16]]


## Gaussian Naive Bayes on Wine Dataset – Summary

### Why this dataset and model?

The Wine dataset from `sklearn.datasets.load_wine` contains 178 wines with 13 continuous chemical features (e.g. alcohol, malic acid, color intensity, proline). These features are numeric and can reasonably be modeled as following a Gaussian (normal) distribution for each class, which matches the core assumption of **Gaussian Naive Bayes** for continuous data.[web:40][web:42][web:63]

Gaussian Naive Bayes treats each feature as conditionally independent given the class and models its distribution using a per‑class mean and variance. It is simple, fast, and well‑suited for small tabular datasets with continuous attributes.[web:54][web:57][web:65]

---

### Baseline model (all 13 features)

I first trained a Gaussian Naive Bayes classifier on all 13 features:

- Train/test split with stratification on the target (`test_size=0.35`, `random_state=42`).  
- Model: `GaussianNB()` from scikit‑learn.  
- Evaluation metrics: accuracy and confusion matrix.

**Result (all features):**

- Accuracy: ~0.95  
- Confusion matrix shows very few misclassifications across the 3 wine classes.

This indicates that when the model has access to the full set of chemical measurements, the per‑class Gaussian distributions separate the three cultivars very well, and even a simple probabilistic classifier achieves strong performance.[web:40][web:42][web:54]

---

### Effect of removing multiple features

Next, I removed several features:

- Dropped: `ash`, `alcalinity_of_ash`, `magnesium`, `hue`, `od280/od315_of_diluted_wines`, and `proline`.  
- Re‑built the train/test split and retrained GaussianNB on the reduced feature set.

**Result (7 features):**

- Accuracy dropped to around **0.87**.  
- The confusion matrix shows more misclassifications, especially between classes 0 and 1.

Interpretation:

Removing multiple chemical attributes reduces how well the Gaussian distributions can distinguish between the classes. The model still works, but with fewer informative features, the likelihoods for different classes become more similar, leading to more overlap and lower accuracy.[web:42][web:53][web:57]

This experiment demonstrates that Gaussian Naive Bayes depends heavily on having enough **informative continuous features** so that the class‑conditional means and variances are meaningfully different.

---

### Standardization after feature removal

To see the effect of feature scaling, I then:

- Dropped one additional feature (`total_phenols`) from the already reduced dataset.  
- Applied `StandardScaler` to the remaining features:
  - Fit the scaler on `X_train`.
  - Transform both `X_train` and `X_test` with the fitted scaler.
- Trained GaussianNB on the scaled data.

**Result (scaled 6‑feature model):**

- Accuracy increased slightly compared to the unscaled reduced‑feature model (e.g. from ~0.87 to ~0.89).  
- The confusion matrix shows a small reduction in misclassifications for one of the classes.

Interpretation:

Gaussian Naive Bayes does not require scaling to work, because it estimates a mean and variance for each feature in each class directly. However, standardization can help numerically and slightly rebalance the influence of features, which sometimes leads to minor improvements. Scaling cannot fully recover the performance lost by removing informative features, but it shows that preprocessing can still matter even for simple probabilistic models.[web:54][web:57]

---

### Key takeaways

- **Dataset–model fit**: Wine is a good choice for Gaussian Naive Bayes because it has continuous numeric features and clear class structure, aligning with the Gaussian assumption.[web:40][web:63]  
- **Importance of informative features**: Removing several chemical attributes noticeably reduces accuracy, confirming that Gaussian NB relies on differences in per‑class means and variances to separate classes.  
- **Role of scaling**: StandardScaler slightly improves performance in the reduced‑feature setting, but the biggest impact on accuracy comes from how many and which features are included, not from scaling alone.[web:54][web:57]

# Overall, this mini‑project illustrates how Gaussian Naive Bayes behaves on a real tabular dataset and how feature selection and preprocessing affect its performance.